### PR TITLE
Adding Multi-Image generation usecase app

### PR DESCRIPTION
**Pull request by ravi9**
## Description

This Streamlit app is designed to generate multiple images based on a provided text prompt. It leverages [TorchServe](https://pytorch.org/serve/) for efficient model serving and management, and utilizes [Meta-LLaMA-3.2](https://huggingface.co/meta-llama) for prompt generation, and **Stable Diffusion** with [latent-consistency/lcm-sdxl](https://huggingface.co/latent-consistency/lcm-sdxl) and [Torch.compile using OpenVINO backend](https://docs.openvino.ai/2024/openvino-workflow/torch-compile.html) for image generation.

### CODE REVIEW

File: examples/usecases/llm_diffusion_serving_app/Readme.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
+
+# Multi-Image Generation App with Streamlit, Llama, Stable Diffusion, OpenVINO, TorchServe
```
Comment (agunapal) [inline examples/usecases/llm_diffusion_serving_app/Readme.md:2]:
```suggestion
# Multi-Image Generation Streamlit App by chaining Llama & Stable Diffusion using TorchServe, torch.compile & OpenVINO
```

File: examples/usecases/llm_diffusion_serving_app/Readme.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+This Streamlit app is designed to generate multiple images based on a provided text prompt. It leverages [TorchServe](https://pytorch.org/serve/) for efficient model serving and management, and utilizes [Meta-LLaMA-3.2](https://huggingface.co/meta-llama) for prompt generation, and **Stable Diffusion** with [latent-consistency/lcm-sdxl](https://huggingface.co/latent-consistency/lcm-sdxl) and [Torch.compile using OpenVINO backend](https://docs.openvino.ai/2024/openvino-workflow/torch-compile.html) for image generation.
```
Comment (agunapal) [inline examples/usecases/llm_diffusion_serving_app/Readme.md:4]: On the second line, please add "Instead of using Stable Diffusion directly, this app chains Llama and Stable Diffusion . It takes a user prompt & makes use of llama to create multiple interesting relevant prompts which are then sent to Stable Diffusion to generate images "

File: examples/usecases/llm_diffusion_serving_app/docker/client_app.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+MODEL_NAME_LLM = os.environ["MODEL_NAME_LLM"]
+MODEL_NAME_LLM = MODEL_NAME_LLM.replace("/", "---")
+
+MODEL_NAME_SD = os.environ["MODEL_NAME_SD"]
+MODEL_NAME_SD = MODEL_NAME_SD.replace("/", "---")
+
+logger = logging.getLogger(__name__)
+
+# App title
+st.set_page_config(page_title="Image Generation with SDXL and OpenVINO")
```
Comment (agunapal) [inline examples/usecases/llm_diffusion_serving_app/docker/client_app.py:23]: The title should include Llama too ?

File: examples/usecases/llm_diffusion_serving_app/docker/client_app.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+
+with st.sidebar:
+    st.title("Image Generation with SDXL and OpenVINO")
```
Comment (agunapal) [inline examples/usecases/llm_diffusion_serving_app/docker/client_app.py:26]: The title should include Llama too ?

File: examples/usecases/llm_diffusion_serving_app/Readme.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  Network URL: http://123.11.0.2:8085
+  External URL: http://123.123.12.34:8085
+
+
+  You can now view your Streamlit app in your browser.
+
+  Local URL: http://localhost:8084
+  Network URL: http://123.11.0.2:8084
+  External URL: http://123.123.12.34:8084
+```
```
Comment (agunapal) [inline examples/usecases/llm_diffusion_serving_app/Readme.md:126]: Can we please include screenshots of what the UI looks like for both server & client, an example of images with prompt

**Review by agunapal — CHANGES_REQUESTED** [review]
Overall looks good to me,

I think the intro section needs more highlighting of what is the differentiated solution here vs using Stable diffusion. There is a lot happening under the hood, which needs to be shown & explained.

Also please check the changes in the docs folder in this PR

https://github.com/pytorch/serve/pull/3276

You need to add this to make it show up in https://pytorch.org/serve/genai_use_cases.html

File: examples/usecases/llm_diffusion_serving_app/docker/sd/stable_diffusion_handler.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+            input_batch (list): List of Text from the pre-process function is passed here
+        Returns:
+            list : It returns a list of the generate images for the input text
+        """
+        # Handling inference for sequence_classification.
+        guidance_scale = model_inputs.get("guidance_scale") or 5.0
+        num_inference_steps = model_inputs.get("num_inference_steps") or 5
+        height = model_inputs.get("height") or 768
+        width = model_inputs.get("width") or 768
+        inferences = self.pipeline(
```
Comment (agunapal) [inline examples/usecases/llm_diffusion_serving_app/docker/sd/stable_diffusion_handler.py:162]: Does the diffusion pipeline set torch.no_grad() ?

File: docs/genai_use_cases.md [hunk 7]
```diff
@@ -5,3 +5,7 @@ This document shows interesting usecases with TorchServe for Gen AI deployments.
 ## [Enhancing LLM Serving with Torch Compiled RAG on AWS Graviton](https://pytorch.org/serve/enhancing_llm_serving_compile_rag.html)
 
 In this blog, we show how to deploy a RAG Endpoint using TorchServe, increase throughput using `torch.compile` and improve the response generated by the Llama Endpoint. We also show how the RAG endpoint can be deployed on CPU using AWS Graviton, while the Llama endpoint is still deployed on a GPU. This kind of microservices-based RAG solution efficiently utilizes compute resources, resulting in potential cost savings for customers.
+
+## [Multi-Image Generation Streamlit App: Chaining LLaMA & Stable Diffusion using TorchServe, torch.compile & OpenVINO](https://pytorch.org/serve/llm_diffusion_serving_app.html)
```
Comment (agunapal) [inline docs/genai_use_cases.md:9]:
```suggestion
## [Multi-Image Generation Streamlit App: Chaining Llama & Stable Diffusion using TorchServe, torch.compile & OpenVINO](https://pytorch.org/serve/llm_diffusion_serving_app.html)
```

File: docs/genai_use_cases.md [hunk 8]
```diff
@@ -5,3 +5,7 @@ This document shows interesting usecases with TorchServe for Gen AI deployments.
...
+
+This Multi-Image Generation Streamlit app is designed to generate multiple images based on a provided text prompt. Instead of using Stable Diffusion directly, this app chains LLaMA and Stable Diffusion to enhance the image generation process. This multi-image generation use case exemplifies the powerful synergy of cutting-edge AI technologies: TorchServe, OpenVINO, Torch.compile, Meta-LLaMA, and Stable Diffusion.
```
Comment (agunapal) [inline docs/genai_use_cases.md:11]:
```suggestion
This Multi-Image Generation Streamlit app is designed to generate multiple images based on a provided text prompt. Instead of using Stable Diffusion directly, this app chains Llama and Stable Diffusion to enhance the image generation process. This multi-image generation use case exemplifies the powerful synergy of cutting-edge AI technologies: TorchServe, OpenVINO, Torch.compile, Meta-Llama, and Stable Diffusion. 
```

File: examples/usecases/llm_diffusion_serving_app/README.md [hunk 9]
```diff
@@ -0,0 +1,213 @@
+
+## Multi-Image Generation Streamlit App: Chaining LLaMA & Stable Diffusion using TorchServe, torch.compile & OpenVINO
+
+This Multi-Image Generation Streamlit app is designed to generate multiple images based on a provided text prompt. Instead of using Stable Diffusion directly, this app chains LLaMA and Stable Diffusion to enhance the image generation process. Here’s how it works:
```
Comment (agunapal) [inline examples/usecases/llm_diffusion_serving_app/README.md:4]:
```suggestion
This Multi-Image Generation Streamlit app is designed to generate multiple images based on a provided text prompt. Instead of using Stable Diffusion directly, this app chains Llama and Stable Diffusion to enhance the image generation process. Here’s how it works:
```

File: examples/usecases/llm_diffusion_serving_app/README.md [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+- The app takes a user prompt and uses [Meta-LLaMA-3.2](https://huggingface.co/meta-llama) to create multiple interesting and relevant prompts.
```
Comment (agunapal) [inline examples/usecases/llm_diffusion_serving_app/README.md:5]:
```suggestion
- The app takes a user prompt and uses [Meta-Llama-3.2](https://huggingface.co/meta-llama) to create multiple interesting and relevant prompts.
```

File: examples/usecases/llm_diffusion_serving_app/docker/client_app.py [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+
+
+# Client Page UI
+intro_container = st.container()
+with intro_container:
+    st.markdown(
+        """
+    ### Multi-Image Generation App with TorchServe and OpenVINO
+    Welcome to the Multi-Image Generation Client App. This app allows you to generate multiple images 
+    from a single text prompt. Simply input your prompt, and the app will enhance it using a LLM (LLaMA) and
```
Comment (agunapal) [inline examples/usecases/llm_diffusion_serving_app/docker/client_app.py:246]:
```suggestion
    from a single text prompt. Simply input your prompt, and the app will enhance it using a LLM (Llama) and 
```

File: examples/usecases/llm_diffusion_serving_app/docker/sd-benchmark.py [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+            
+    headers = ["Run Mode", "Warm-up Time", f"Average Time for {num_iter} iter", "Image Saved as "]
+    print(tabulate(table_data, headers=headers, tablefmt="grid"))
+    
+    # Save results
+    save_results(results, hw_config, sw_versions)
+    print(f"\nResults and Images saved at {out_dir} which is a Docker container mount, corresponds to 'serve/model-store-local/' on the host machine.\n")
+    
+if __name__ == "__main__":
+    main()
```
Comment (agunapal) [inline examples/usecases/llm_diffusion_serving_app/docker/sd-benchmark.py:278]: Very cool and comprehensive script for benchmarking!

File: examples/usecases/llm_diffusion_serving_app/docker/sd/stable_diffusion_handler.py [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+        else:
+            pipe = StableDiffusionPipeline.from_pretrained(
+                ckpt, torch_dtype=dtype, use_safetensors=True, safety_checker=None
+            )
+
+        if compile_unet:
+            logger.info("Compiling UNet.")
+            if compile_mode == "max-autotune" and change_comp_config:
+                pipe.unet.to(memory_format=torch.channels_last)
+                torch._inductor.config.conv_1x1_as_mm = True
```
Comment (agunapal) [inline examples/usecases/llm_diffusion_serving_app/docker/sd/stable_diffusion_handler.py:97]: What are these configs? Are these missing in the benchmarking script?

File: examples/usecases/llm_diffusion_serving_app/README.md [hunk 14]
```diff
@@ -0,0 +1,213 @@
+
+## Multi-Image Generation Streamlit App: Chaining LLaMA & Stable Diffusion using TorchServe, torch.compile & OpenVINO
```
Comment (agunapal) [inline examples/usecases/llm_diffusion_serving_app/README.md:2]:
```suggestion
## Multi-Image Generation Streamlit App: Chaining Llama & Stable Diffusion using TorchServe, torch.compile & OpenVINO
```

File: examples/usecases/llm_diffusion_serving_app/docker/client_app.py [hunk 15]
```diff
@@ -0,0 +1,355 @@
...
+
+# Initialize session state if not already set
+st.session_state.models_loaded = st.session_state.get("models_loaded", False)
+st.session_state.gen_images = st.session_state.get("gen_images", [])
+st.session_state.gen_captions = st.session_state.get("gen_captions", [])
+st.session_state.llm_prompts = st.session_state.get("llm_prompts", None)
+st.session_state.llm_time = st.session_state.get("llm_time", 0)
+
+with st.sidebar:
+    st.title("Image Generation with LLaMA, SDXL and OpenVINO")
```
Comment (agunapal) [inline examples/usecases/llm_diffusion_serving_app/docker/client_app.py:32]:
```suggestion
    st.title("Image Generation with Llama, SDXL, torch.compile and OpenVINO")
```

**Review by agunapal — CHANGES_REQUESTED** [review]
Looks great overall. minor edits needed for Llama vs LLaMA

File: examples/usecases/llm_diffusion_serving_app/docker/sd/stable_diffusion_handler.py (hunk 6 above)
Comment (ravi9) [reply examples/usecases/llm_diffusion_serving_app/docker/sd/stable_diffusion_handler.py:162]: Yes, it does.
https://huggingface.co/docs/diffusers/main/en/api/pipelines/overview

File: examples/usecases/llm_diffusion_serving_app/docker/sd/stable_diffusion_handler.py (hunk 13 above)
Comment (agunapal) [reply examples/usecases/llm_diffusion_serving_app/docker/sd/stable_diffusion_handler.py:97]: Oh I see. In the blog , we can add a note about these configs for GPU and what benefit they provide

File: examples/usecases/llm_diffusion_serving_app/docker/sd/stable_diffusion_handler.py (hunk 6 above)
Comment (agunapal) [reply examples/usecases/llm_diffusion_serving_app/docker/sd/stable_diffusion_handler.py:162]: Something that I have observed is that for compile, torch.inference_mode() gives better perf compared to torch.no_grad(). We don't need to address this in this PR

Comment (ravi9) [reply examples/usecases/llm_diffusion_serving_app/docker/sd/stable_diffusion_handler.py:162]: Thanks for letting me know. Yes, I'm observing the same when I try to profile, it shows a different performance ! As with the profiler, it uses torch.inference_mode() !!

**Review by agunapal — APPROVED** [review]
LGTM
